src/cli.rs

```diff
@@ -1,49 +1,139 @@
-use clap::*;
-use std::ffi::OsString;
+use crate::error::PnError;
+use crate::{commands, commands::CommandTrait, error::MainError};
+use clap::{Parser, Subcommand};
+use itertools::Itertools;
+use pipe_trait::Pipe;
+use std::{env, ffi::OsString, num::NonZeroI32, path::Path, process, process::Stdio};
 
 #[derive(Debug, Parser)]
 #[clap(author, version, about, rename_all = "kebab-case")]
 pub struct Cli {
-    /// Run the command on the root workspace project.
-    #[clap(short, long)]
-    pub workspace_root: bool,
+    #[clap(flatten)]
+    pub config: Config,
+
     /// Command to execute.
     #[clap(subcommand)]
     pub command: Command,
 }
 
+/// Global configurations to be passed down to subcommands
+#[derive(Parser, Debug)]
+pub struct Config {
+    /// Run the command on the root workspace project
+    #[clap(short, long)]
+    pub workspace_root: bool,
+}
+
 #[derive(Debug, Subcommand)]
 #[clap(rename_all = "kebab-case")]
 pub enum Command {
     /// Runs a defined package script.
     #[clap(alias = "run-script")]
-    Run(RunArgs),
+    Run(commands::run::Run),
+
     /// Installs all dependencies of the project in the current working directory.
     /// When executed inside a workspace, installs all dependencies of all projects.
     #[clap(alias = "i")]
-    Install(PassedThroughArgs),
+    Install(commands::PassedThroughArgs),
+
     /// Updates packages to their latest version based on the specified range.
     /// You can use "*" in package name to update all packages with the same pattern.
     #[clap(alias = "up")]
-    Update(PassedThroughArgs),
+    Update(commands::PassedThroughArgs),
+
     /// Execute a shell command in scope of a project.
     #[clap(external_subcommand)]
     Other(Vec<String>),
 }
 
-#[derive(Debug, Args)]
-#[clap(rename_all = "kebab-case")]
-pub struct PassedThroughArgs {
-    pub args: Vec<OsString>,
+impl Cli {
+    pub fn run(self) -> Result<(), MainError> {
+        match self.command {
+            Command::Run(e) => e.run(self.config),
```

Comment:
So you defined a trait only to use it once? I suggest you remove the trait.

Reply:
i added the trait cz i thought thered be more commands added later on or something. ig for now it seems useless. so ill remove it then.